Remove hypothesis.extra.django's dependence on faker

DRMacIver opened the pull request:
This creates a small internal use only email strategy that replaced hypothesis.extra.django's use of the faker based email strategy.

The strategy in question is not very good, but it is mostly better than the fake factory based one (the big limitation in comparison is that the email addresses generated only contain ascii, but I think that's probably a feature rather than a bug given handling of unicode emails).

This is partly an experiment. I want to see if people mention the fake factory integration ever again when the Django integration stops needing it. The faker integration is also really slow right now, so moving it out of the critical path for testing Django is a nice bonus.

alexwlchan commented on scripts/basic-test.sh:
```diff
@@ -46,15 +46,16 @@ if [ "$DARWIN" = true ]; then
 
 # fake-factory doesn't have a correct universal wheel
-pip install --no-use-wheel .[fakefactory]
+pip install --no-use-wheel faker
```
I’m not sure what version of pip runs on Travis, but when I run this command with pip 9.0.1, I get a deprecation warning:

```
DEPRECATION: --no-use-wheel is deprecated and will be removed in the future.  Please use --no-binary :all: instead.
```

Perhaps worth tweaking the command while we’re here?

DRMacIver replied:
I get the same, and thought about changing it, but I kinda want to be better about one fix per PR.

IOU one additional PR after this?

alexwlchan replied:
Fine for me.

alexwlchan commented on src/hypothesis/extra/django/models.py:
```diff
@@ -95,6 +95,33 @@ def validate(value):
+
+domain = st.builds(
+    lambda x, y: '.'.join(x + [y]),
+    st.lists(st.text(safe_letters, min_size=1)), st.sampled_from([
```
This strategy will sometimes produce bare TLDs, and regularly. That seems like a bug. e.g.

```python
for _ in range(25):
    d = domain.example()
    if '.' not in d:
        print(d)
```

returns 

```
com
net
net
com
com
biz
```

Running a similar test with the old strategy doesn’t turn up any email addresses that didn’t have a multi-part domain name after the `@` symbol:

```python
for _ in range(1000):
    e = ff(u'email').example()
    if '.' not in e.split('@')[1]:
        print(e)
```

doesn’t return anything for me.

Should there be a `min_size=1` for `st.lists()` as well?

DRMacIver replied:
Yes, good catch. There definitely should be.

DRMacIver commented:
> And so the drive for zero dependencies continues…

We Have Always Been At War With Dependencies

The faker dependency has a history of being particularly problematic though, and I would like to remove that entire extra package in the medium to long run.

> I’d like to change the domain strategy so it doesn’t return bare TLDs (or be convinced that the current behaviour is sensible).

I agree.

> There aren’t any tests on this patch. Should there be, or is this code sufficiently covered by the existing Django tests?

It's complicated. These are definitely exercised by the existing tests to the degree that the faker code was (we'd see coverage errors for those lambdas if they weren't I think!). I suspect neither are currently passing Django's validators for emails (I *know* the faker based ones aren't), but more work is needed to get things properly validating than just this.

> More generally: I assume the intermediate strategies don’t become part of the public API immediately?

Correct.

> Long term, should they be moved into Hypothesis core? I can imagine a strategy for email addresses and domain names might be useful to non-Django users.

They should, but these are sufficiently half-arsed right now that I don't want to make them more widely available. They satisfy the criterion of "better than using faker" and that's about it.

> (And based on a cursory glance at the faker code, it looks like this version shrinks a lot better than the faker-based code.)

That should be the case, yes. That's one of the reasons I want to drop the faker integration.

DRMacIver commented:
The fact that I allowed to generated bare TLDs convinced me I should do at least *some* validation tests, so I added a very small set (these aren't really intended to be sufficient to cover all of the behaviour you'd expect but are complementary to the existing tests that you can create users with these emails).

I've also fixed that problem and made the naming more consistently pluralised.

alexwlchan commented:
CI failure seems to be unrelated to the changes, so I’ve restarted the job.

Otherwise this all looks good. 👍
